cmake/KokkosConfigCommon.cmake.in

```diff
@@ -203,6 +204,573 @@ int main()
+    list(APPEND INP_DIRECTORY ${CMAKE_SOURCE_DIR})
+    unset(INP_GLOBAL)
+  endif()
+
+  #set properties on directories
+  foreach(directory IN LISTS INP_DIRECTORY)
+    kokkos_get_directory_targets(
+      FOUND_TARGETS DIRECTORY_TARGETS SEARCH_DIR ${directory} EXCLUDE_DIRS ${Kokkos_EXCLUDE_BUILD_DIRS}
+    )
+
+    foreach(directory_target IN LISTS DIRECTORY_TARGETS)
+      kokkos_check_compile_dependency(
+        compile_dependency
+        transitive_compile_dependency
+        link_dependency
+        "Kokkos::kokkoscore;Kokkos::kokkoscontainers;Kokkos::kokkosalgorithms;Kokkos::kokkossimd;Kokkos::kokkos"
+        ${directory_target}
+        KOKKOS_LINK_DEPENDENCY_CACHE
+        KOKKOS_PUBLIC_DEPENDENCY_CACHE
+      )
+      if(compile_dependency)
+        list(APPEND DIRECTORY_COMPILE_TARGETS ${directory_target})
+        list(APPEND COMPILE_TARGETS ${directory_target})
+      endif()
+
+      if(link_dependency)
+        list(APPEND LINK_TARGETS ${directory_target})
+      endif()
+    endforeach()
+
+    # set all cpp,cxx sources in the found targets to the correct language
+    if(@Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE@)
+      foreach(target IN LISTS DIRECTORY_COMPILE_TARGETS)
+        get_target_property(target_sources ${target} SOURCES)
+        get_target_property(target_src_dir ${target} SOURCE_DIR)
+        get_target_property(target_bin_dir ${target} BINARY_DIR)
+        foreach(source IN LISTS target_sources)
+          if(source)
+            cmake_path(ABSOLUTE_PATH source BASE_DIRECTORY ${target_src_dir} OUTPUT_VARIABLE source_abs_path)
+            if(NOT EXISTS ${source_abs_path})
+              message(
+                DEBUG
+                "Kokkos assumes that file ${source} for target ${target} is a generated file written into the binary directory ${target_bin_dir}"
+              )
+              cmake_path(ABSOLUTE_PATH source BASE_DIRECTORY ${target_bin_dir} OUTPUT_VARIABLE source_abs_path)
+              if(NOT EXISTS ${source_abs_path})
+                message(
+                  FATAL_ERROR
+                    "Kokkos could not find the file ${source} for target ${target} at location ${source_abs_path}."
+                )
+              endif()
+            endif()
+            cmake_path(GET source_abs_path EXTENSION LAST_ONLY extension)
+            string(REGEX REPLACE "^\\." "" extension_no_dot "${extension}")
+            if(${extension_no_dot} IN_LIST CMAKE_CXX_SOURCE_FILE_EXTENSIONS)
+              message(
+                DEBUG
+                "Kokkos setting language to @KOKKOS_COMPILE_LANGUAGE@ on target ${target} for source ${source_abs_path}"
+              )
+              set_property(
+                SOURCE ${source_abs_path} TARGET_DIRECTORY ${target} PROPERTY LANGUAGE @KOKKOS_COMPILE_LANGUAGE@
+              )
+            endif()
+          endif()
+        endforeach()
+      endforeach()
+    endif()
```

Comment:
Can this thing be put in a function? It looks like a duplicate of L488.

Reply:
Since this will be executed as part of a deferred call it has to be visible in the scope that the deferred call is dispatched in. The easiest solution for this was to have it in both the respective functions

Reply:
When calling functions in a deferred call, the function that should be executed has to be visible in the scope of the deferred call (the end of the CMakeLists.txt that the call gets actually run). Since we defer a call to `CMAKE_SOURCE_DIR` we are in the highest scope. With this it was easiest to just repeat here even if I agree with you and not really like the repetition.